syntax = "proto3";

package chat;

service ChatService {
    // Unary RPC: User login
    rpc Login(LoginRequest) returns (LoginResponse);

    // Server Streaming: Get recent messages
    rpc GetRecentMessages(Empty) returns (stream MessageResponse);

    // Client Streaming: Broadcast chat messages
    rpc BroadcastChat(stream MessageRequest) returns (BroadcastResponse);

    // Bidirectional Streaming: Real-time chat
    rpc ChatStream(stream MessageRequest) returns (stream MessageResponse);
}

message LoginRequest {
    string username = 1;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
}

message Empty {}

message MessageRequest {
    string username = 1;
    string message = 2;
}

message MessageResponse {
    string username = 1;
    string message = 2;
}

message BroadcastResponse {
    int32 count = 1;
}
